src/OpenTelemetry/Metrics/HistogramBuckets.cs

```diff
@@ -104,5 +186,18 @@ public bool MoveNext()
                 return false;
             }
         }
+
+        private sealed class Bucket
+        {
+            public double UpperBoundInclusive { get; set; }
+
+            public double LowerBoundExclusive { get; set; }
```

Comment:
Do we need to track two values for each node?

I think we should able to search for a value by inspecting the left node's value instead of tracking a LowerBoundExclusive for each node of the tree here:
```csharp
if (value <= current.LowerBoundExclusive)
{
    current = current.Left;
}
```

Reply:
Define "need to" 😄 Technically, no. But that is an optimization to make it fast[er]. In order to remove that we would need to (do something like) track the previous node.

I have the benchmarks on [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3252/files#r912263407) if you want to mess with it and see. Not opposed to making changes, just not sure how to do it without slowing it down.

Basically the running assumption is that executing the binary search is critical/hot-path (executed any time a measurement is taken) so trading 8 bytes of memory on init to improve that per-call perf is worth it.

Reply:
I tried messing around with the benchmarks code and I found a way to do this without the using additional field but it slower than the current approach by ~7ns. I'm not sure if that's good enough though:

// * Summary *

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Core i7-9700 CPU 3.00GHz, 1 CPU, 8 logical and 8 physical cores
.NET SDK=6.0.301
  [Host]     : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT
  DefaultJob : .NET 6.0.6 (6.0.622.26707), X64 RyuJIT


|                                      Method | BoundCount |     Mean |    Error |   StdDev |
|-------------------------------------------- |----------- |---------:|---------:|---------:|
|                   FindBucketIndexBinaryTree |       1000 | 56.43 ns | 0.568 ns | 0.504 ns |
| FindBucketIndexBinaryTreeWithoutExtraMemory |       1000 | 63.60 ns | 0.870 ns | 0.771 ns |

<details>

<summary>Expand for code</summary>

```csharp
// <copyright file="HistogramBenchmarks.cs" company="OpenTelemetry Authors">
// Copyright The OpenTelemetry Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>

using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;

namespace Benchmarks.Metrics
{
    public class HistogramBenchmarks
    {
        private const int MaxValue = 10000;
        private readonly Random random = new();
        private double[] bounds;
        private Bucket root;

        [Params(1000)]
        public int BoundCount { get; set; }

        [GlobalSetup]
        public void Setup()
        {
            // Evenly distribute the bound values over the range [0, MaxValue)
            this.bounds = new double[this.BoundCount];
            for (int i = 0; i < this.bounds.Length; i++)
            {
                this.bounds[i] = i * MaxValue / this.bounds.Length;
            }

            this.root = ConstructBalancedTree(this.bounds, 0, this.bounds.Length);

            static Bucket ConstructBalancedTree(double[] values, int min, int max)
            {
                if (min == max)
                {
                    return null;
                }

                int median = min + ((max - min) / 2);
                return new Bucket
                {
                    Index = median,
                    UpperBoundInclusive = values[median],
                    LowerBoundExclusive = median > 0 ? values[median - 1] : double.MinValue,
                    Left = ConstructBalancedTree(values, min, median),
                    Right = ConstructBalancedTree(values, median + 1, max),
                };
            }
        }

        [Benchmark]
        public int FindBucketIndexBinaryTree()
        {
            return this.FindBucketIndexBinaryTree(this.random.Next(MaxValue));
        }

        [Benchmark]
        public int FindBucketIndexBinaryTreeWithoutExtraMemory()
        {
            return this.FindBucketIndexBinaryTreeWithoutExtraMemory(this.random.Next(MaxValue));
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexBinaryTree(double value)
        {
            Bucket current = this.root;

            Debug.Assert(current != null, "Bucket root was null.");

            do
            {
                if (value <= current.LowerBoundExclusive)
                {
                    current = current.Left;
                }
                else if (value > current.UpperBoundInclusive)
                {
                    current = current.Right;
                }
                else
                {
                    return current.Index;
                }
            }
            while (current != null);

            return this.bounds.Length;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int FindBucketIndexBinaryTreeWithoutExtraMemory(double value)
        {
            Bucket current = this.root;

            Debug.Assert(current != null, "Bucket root was null.");

            while (true)
            {
                if (value <= current.UpperBoundInclusive)
                {
                    if (current.Left == null)
                    {
                        return current.Index;
                    }

                    current = current.Left;
                }
                else
                {
                    if (current.Right == null)
                    {
                        break;
                    }

                    current = current.Right;
                }
            }

            return this.bounds.Length;
        }

        private sealed class Bucket
        {
            public double UpperBoundInclusive { get; set; }

            public double LowerBoundExclusive { get; set; }

            public int Index { get; set; }

            public Bucket Left { get; set; }

            public Bucket Right { get; set; }
        }
    }
}
```
</details>

Reply:
I guess it's okay to have an additional 8 bytes of memory for those 7 extra nanoseconds that we would gain for every measurement?